File: pages/api/chat-analysis.ts
import { NextApiRequest, NextApiResponse } from 'next';
import { openai } from '../../lib/openaiClient';
import { supabase } from '../../lib/supabaseClient';

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    const { startupId, message, conversationHistory = [] } = req.body;
    if (!startupId || !message) {
      return res.status(400).json({ error: 'Startup ID and message are required' });
    }

    // 1. Fetch startup context
    const { data: startup, error: startupError } = await supabase
      .from('startups')
      .select('*')
      .eq('id', startupId)
      .single();

    if (startupError || !startup) {
      return res.status(404).json({ error: 'Startup not found' });
    }

    // 2. Fetch recent documents and analysis
    const { data: documents } = await supabase
      .from('documents')
      .select('name, type, summary, kpis, red_flags')
      .eq('startup_id', startupId)
      .limit(5);

    const { data: metrics } = await supabase
      .from('metrics')
      .select('*')
      .eq('startup_id', startupId)
      .single();

    const { data: notes } = await supabase
      .from('notes')
      .select('content, created_at')
      .eq('startup_id', startupId)
      .order('created_at', { ascending: false })
      .limit(3);

    // 3. Prepare context for AI
    const context = `
CONTEXTO DE LA STARTUP:
- Nombre: ${startup.name}
- Sector: ${startup.sector}
- País: ${startup.country}
- Fase: ${startup.stage}
- Descripción: ${startup.description || 'No disponible'}

DOCUMENTOS RECIENTES:
${documents?.map(doc => 
  `- ${doc.name} (${doc.type}): ${doc.summary || 'Sin resumen'}`
).join('\n') || 'No hay documentos'}

MÉTRICAS ACTUALES:
${metrics ? 
  `ARR: ${metrics.arr}, MRR: ${metrics.mrr}, CAC: ${metrics.cac}, LTV: ${metrics.ltv}, Churn: ${metrics.churn}, Runway: ${metrics.runway} meses` : 
  'No hay métricas disponibles'}

NOTAS RECIENTES:
${notes?.map(note => 
  `- ${new Date(note.created_at).toLocaleDateString()}: ${note.content}`
).join('\n') || 'No hay notas'}

HISTORIAL DE CONVERSACIÓN:
${conversationHistory.map((msg: any) => 
  `${msg.role}: ${msg.content}`
).join('\n')}

USUARIO: ${message}

INSTRUCCIONES:
Eres un asistente experto en análisis de inversiones para venture capital. Tu objetivo es ayudar a analizar esta startup y mantener conversaciones productivas sobre:

1. Análisis de inversión (pros/contras, riesgos, oportunidades)
2. Actualización de memos de inversión
3. Research de mercado y competidores
4. Preguntas para due diligence
5. Comparación con otras startups del portfolio
6. Recomendaciones de seguimiento

Responde de forma clara, estructurada y útil. Si el usuario pide actualizar el memo de inversión, proporciona el texto actualizado completo.
`;

    const completion = await openai.createCompletion({
      model: 'text-davinci-003',
      prompt: context,
      max_tokens: 1000,
      temperature: 0.7,
    });

    const response = completion.data.choices[0]?.text?.trim();
    if (!response) {
      return res.status(500).json({ error: 'No response generated' });
    }

    res.status(200).json({ 
      response,
      startupName: startup.name,
      timestamp: new Date().toISOString()
    });

  } catch (error) {
    console.error('Error in chat analysis:', error);
    res.status(500).json({ error: 'Error in chat analysis' });
  }
} 

File: backend/routers/chat_analysis.py
from fastapi import APIRouter, HTTPException
from datetime import datetime

from schemas import ChatAnalysisRequest, ChatAnalysisResponse, ErrorResponse
from clients import openai_client, get_startup_by_id, get_documents_by_startup_id, get_metrics_by_startup_id, get_notes_by_startup_id

router = APIRouter()

@router.post("/chat-analysis", response_model=ChatAnalysisResponse)
async def chat_analysis(request: ChatAnalysisRequest):
    """
    Interactive chat analysis for startup insights
    """
    try:
        # 1. Fetch startup context
        startup = await get_startup_by_id(request.startup_id)
        if not startup:
            raise HTTPException(status_code=404, detail="Startup not found")

        # 2. Fetch recent documents and analysis (limit for context)
        documents = await get_documents_by_startup_id(request.startup_id)
        if documents and len(documents) > 5:
            documents = documents[:5]  # Limit to 5 most recent

        metrics = await get_metrics_by_startup_id(request.startup_id)
        notes = await get_notes_by_startup_id(request.startup_id, limit=3)

        # 3. Prepare context for AI
        documents_text = "No hay documentos"
        if documents:
            documents_text = "\n".join([
                f"- {doc.get('name', 'N/A')} ({doc.get('type', 'N/A')}): {doc.get('summary', 'Sin resumen')}"
                for doc in documents
            ])

        metrics_text = "No hay métricas disponibles"
        if metrics:
            metrics_text = f"ARR: {metrics.get('arr', 'N/A')}, MRR: {metrics.get('mrr', 'N/A')}, CAC: {metrics.get('cac', 'N/A')}, LTV: {metrics.get('ltv', 'N/A')}, Churn: {metrics.get('churn', 'N/A')}, Runway: {metrics.get('runway', 'N/A')} meses"

        notes_text = "No hay notas"
        if notes:
            notes_text = "\n".join([
                f"- {datetime.fromisoformat(note['created_at'].replace('Z', '+00:00')).strftime('%Y-%m-%d')}: {note['content']}"
                for note in notes
            ])

        conversation_history_text = ""
        if request.conversation_history:
            conversation_history_text = "\n".join([
                f"{msg.get('role', 'user')}: {msg.get('content', '')}"
                for msg in request.conversation_history
            ])

        context = f"""
CONTEXTO DE LA STARTUP:
- Nombre: {startup.get('name', 'N/A')}
- Sector: {startup.get('sector', 'N/A')}
- País: {startup.get('country', 'N/A')}
- Fase: {startup.get('stage', 'N/A')}
- Descripción: {startup.get('description', 'No disponible')}

DOCUMENTOS RECIENTES:
{documents_text}

MÉTRICAS ACTUALES:
{metrics_text}

NOTAS RECIENTES:
{notes_text}

HISTORIAL DE CONVERSACIÓN:
{conversation_history_text}

USUARIO: {request.message}

INSTRUCCIONES:
Eres un asistente experto en análisis de inversiones para venture capital. Tu objetivo es ayudar a analizar esta startup y mantener conversaciones productivas sobre:

1. Análisis de inversión (pros/contras, riesgos, oportunidades)
2. Actualización de memos de inversión
3. Research de mercado y competidores
4. Preguntas para due diligence
5. Comparación con otras startups del portfolio
6. Recomendaciones de seguimiento

Responde de forma clara, estructurada y útil. Si el usuario pide actualizar el memo de inversión, proporciona el texto actualizado completo.
"""

        # Generate response using OpenAI
        response = openai_client.chat.completions.create(
            model="gpt-3.5-turbo",
            messages=[
                {"role": "system", "content": "Eres un asistente experto en análisis de inversiones para venture capital."},
                {"role": "user", "content": context}
            ],
            max_tokens=1000,
            temperature=0.7
        )

        ai_response = response.choices[0].message.content.strip()
        if not ai_response:
            raise HTTPException(status_code=500, detail="No response generated")

        return ChatAnalysisResponse(
            response=ai_response,
            startup_name=startup['name'],
            timestamp=datetime.utcnow()
        )

    except HTTPException:
        raise
    except Exception as e:
        print(f"Error in chat analysis: {e}")
        raise HTTPException(status_code=500, detail="Error in chat analysis")